tutorials/sphinx-tutorials/multiagent_ppo.py

```diff
@@ -0,0 +1,789 @@
+        share_params=share_parameters_policy,
+        device=device,
+        depth=2,
+        num_cells=256,
+        activation_class=torch.nn.Tanh,
+    ),
+    NormalParamExtractor(),  # this will just separate the last dimension into two outputs: a loc and a non-negative scale
+)
+
+######################################################################
+# **Second**: wrap the neural network in a :class:`TensorDictModule`
+#
+# This is simply a module that will read the ``in_keys`` from a tensordict, feed them to the
+# neural networks, and write the
+# outputs in-place at the ``out_keys``.
+#
+# Note that we use ``("agents", ...)`` keys as these keys are denoting data with the
+# additional ``n_agents`` dimension.
+#
+policy_module = TensorDictModule(
+    policy_net,
+    in_keys=[("agents", "observation")],
+    out_keys=[("agents", "loc"), ("agents", "scale")],
+)
+
+######################################################################
+# **Third**: wrap the :class:`TensorDictModule` in a :class:`ProbabilisticActor`
+#
+# We now need to build a distribution out of the location and scale of our
+# normal distribution. To do so, we instruct the :class:`ProbabilisticActor`
+# class to build a :class:`TanhNormal` out of the location and scale
+# parameters. We also provide the minimum and maximum values of this
+# distribution, which we gather from the environment specs.
+#
+# The name of the ``in_keys`` (and hence the name of the ``out_keys`` from
+# the :class:`TensorDictModule` above) has to end with the
+# :class:`TanhNormal` distribution constructor keyword arguments (loc and scale).
+#
+
+policy = ProbabilisticActor(
+    module=policy_module,
+    spec=env.unbatched_action_spec,
+    in_keys=[("agents", "loc"), ("agents", "scale")],
+    out_keys=[env.action_key],
+    distribution_class=TanhNormal,
+    distribution_kwargs={
+        "min": env.unbatched_action_spec[env.action_key].space.minimum,
+        "max": env.unbatched_action_spec[env.action_key].space.maximum,
+    },
+    return_log_prob=True,
+    log_prob_key=("agents", "sample_log_prob"),
+)  # we'll need the log-prob for the PPO loss
+
+######################################################################
+# Critic network
+# --------------
+#
+# The critic network is a crucial component of the PPO algorithm, even though it
+# isn't used at sampling time. This module will read the observations and
+# return the corresponding value.
+#
+# As before, you have to think carefully about the decision of **sharing the critic parameters**.
+# Sharing is definitely not recommended when agents have different reward functions. In cases
+# where the reward function (note: the reward function, NOT the reward) is the same for all agents (like here),
+# sharing might help performance. In general, even with the same reward function, this choice might have important
+# impacts you need to consider when designing your networks.
```

Comment:
```suggestion
# impact that one should consider when designing your networks.
```

Reply:
This discussion about sharing parameters or not feels a bit unfinished. We never say what is the drawback (except that you have less degrees of freedom?)
IDK if there are theoretical results.
A good conclusion could be something along the line of "The best way to know which choice is preferable is to quickly experiment both options."

Reply:
i have updated